backend/clubs/views.py

```diff
@@ -5056,6 +5058,76 @@ def qr(self, request, *args, **kwargs):
         qr_image.save(response, "PNG")
         return response
 
+    @action(detail=True, methods=["post"])
+    @transaction.atomic
+    def transfer(self, request, *args, **kwargs):
+        """
+        Transfer a ticket to another user
+        ---
+        requestBody:
+            content:
+                application/json:
+                    schema:
+                        type: object
+                        properties:
+                            username:
+                                type: string
+                        required:
+                            - username
+        responses:
+            "200":
+                content:
+                    application/json:
+                        schema:
+                           type: object
+                           properties:
+                                detail:
+                                    type: string
+            "403":
+                content:
+                    application/json:
+                        schema:
+                           type: object
+                           properties:
+                                detail:
+                                    type: string
+            "404":
+                content:
+                    application/json:
+                        schema:
+                           type: object
+                           properties:
+                                detail:
+                                    type: string
+        ---
+        """
+        receiver = get_object_or_404(
+            get_user_model(), username=request.data.get("username")
+        )
+
+        # checking whether the request's user owns the ticket is handled by the queryset
+        ticket = self.get_object()
+        if not ticket.transferrable:
+            return Response(
+                {"detail": "The ticket is non-transferrable"},
+                status=status.HTTP_403_FORBIDDEN,
+            )
+
+        if self.request.user == receiver:
+            return Response(
+                {"detail": "You cannot transfer a ticket to yourself"},
+                status=status.HTTP_403_FORBIDDEN,
+            )
+
+        ticket.owner = receiver
+        ticket.save()
+        TicketTransferRecord.objects.create(
+            ticket=ticket, sender=self.request.user, receiver=receiver
+        ).send_confirmation_email()
```

Comment:
At this point, you'll probably need to invalidate the sender's QR, currently built using

`url = f"https://{settings.DOMAIN}/api/tickets/{self.id}"`

This URL was put in as a placeholder and is not actually suitable. We want a URL that is unique to the user/locked behind auth, so that others can't access it. We could use a [hash of ticket.UUID + ticket.owner](https://en.wikipedia.org/wiki/Object-capability_model), since locking behind auth means locking behind Penn access. If we change the way we do the URL, we would need to write another endpoint that validates it.

@julianweng could we co-ordinate with Anthony on the quickest way to get ticket QR code validation into iOS? Or do we want to use a different interface?

Reply:
After some discussion I think we're going to put this in a separate PR. I've made the updates locally but more changes will likely be required until we finalize details about QR code scanning.